CONTRIBUTING.md

```diff
@@ -1,16 +1,721 @@
+
+#### Output files
+
+FAST stages rely on generated provider and tfvars files, as an easy convenience that allows automated setup and passing of contract values between stages.
+
+Files are written to a special GCS bucket in order to be leveraged by both humans and CI/CD workflows, and optionally also written to local storage if needed.
+
+When editing or adding a stage, you are expected to maintain the output files system so any new contact output is also present in files.
+
+### Style guide reference
+
+Similarly to our design principles above, we evolved a set of style conventions that we try to standardize on to make code more legible and uniform. This reduces friction when coding, and ideally moves us closer to the goal of using IaC as live documentation.
+
+#### Group logical resources or modules in separate files
+
+Over time and as our codebase got larger, we switched away from the canonical `main.tf`/`outputs.tf`/`variables.tf` triplet of file names and now tend to prefer descriptive file names that refer to the logical entities (resources or modules) they contain.
+
+We still use traditional names for variables and outputs, but tend to use main only for top-level locals or resources (e.g. the project resource in the `project` module), or for those resources that would end up in very small files.
+
+While some older modules and examples are still using three files, we are slowly bringing all code up to date and any new development should use descriptive file names.
+
+Our `tfdoc` tool has a way of generating a documentation table that maps file names with descriptions and the actual resources and modules they contain, refer to the last section for details on how to activate the mode in your code.
+
+#### Enforce line lengths
+
+We enforce line length for legibility, and adopted the 79 characters convention from other languages for simplicity.
+
+This convention is relaxed for long resource attribute names (even though in some cases you might want to alias them to short local names), and for variable and output descriptions.
+
+In most other cases you should break long lines, especially in `for` and `for_each` loops. Some of the conventions we adopted:
+
+- break after opening and before closing braces/parenthesis
+- break after a colon in `for` loops
+- add extra parenthesis and breaks to split long ternary operators
+- break right before the `:` and `?` in long ternary operators
+
+This is one of many examples.
+
+```hcl
+locals {
+  sink_bindings = {
+    for type in ["bigquery", "pubsub", "logging", "storage"] :
+    type => {
+      for name, sink in var.logging_sinks :
+      name => sink if sink.iam && sink.type == type
+    }
+  }
+}
+```
+
+#### Use alphabetical order for locals, outputs, variables
```

Comment:
Don't agree for locals. IMO sometimes logically grouping local is easier for readbility.